src/metrics.ts

```diff
@@ -180,107 +194,124 @@ export default class Metrics extends EventEmitter {
+        this.reAddBucket(payload.bucket);
         this.emit(UnleashEvents.Warn, `${url} returning ${res.status}`, await res.text());
       } else {
         this.emit(UnleashEvents.Sent, payload);
       }
     } catch (err) {
-      this.emit(UnleashEvents.Error, err);
+      this.reAddBucket(payload.bucket);
+      this.emit(UnleashEvents.Warn, err);
       this.startTimer();
     }
     return true;
   }
 
   assertBucket(name: string) {
-    if (this.disabled || !this.bucket) {
+    if (this.disabled) {
       return false;
     }
     if (!this.bucket.toggles[name]) {
       this.bucket.toggles[name] = {
         yes: 0,
         no: 0,
+        variants: {},
       };
     }
     return true;
   }
 
   count(name: string, enabled: boolean): boolean {
-    if (this.disabled || !this.bucket) {
+    if (this.disabled) {
       return false;
     }
-    this.assertBucket(name);
-    this.bucket.toggles[name][enabled ? 'yes' : 'no']++;
+    this.increaseCounter(name, enabled, 1);
     this.emit(UnleashEvents.Count, name, enabled);
     return true;
   }
 
   countVariant(name: string, variantName: string) {
-    if (this.disabled || !this.bucket) {
+    if (this.disabled) {
       return false;
     }
-    this.assertBucket(name);
-    const { variants } = this.bucket.toggles[name];
-    if (typeof variants !== 'undefined') {
-      if (!variants[variantName]) {
-        variants[variantName] = 1;
-      } else {
-        variants[variantName]++;
-      }
-    } else {
-      this.bucket.toggles[name].variants = {
-        [variantName]: 1,
-      };
-    }
+    this.increaseVariantCounter(name, variantName, 1);
 
     this.emit(UnleashEvents.CountVariant, name, variantName);
     return true;
   }
 
-  private bucketIsEmpty() {
-    if (!this.bucket) {
+  private increaseCounter(name: string, enabled: boolean, inc = 1): boolean {
+    if(inc === 0) {
       return false;
     }
+    this.assertBucket(name);
+    this.bucket.toggles[name][enabled ? 'yes' : 'no'] += inc;
+    return true;
+  }
+
+  private increaseVariantCounter(name: string, variantName: string, inc = 1): boolean {
+    this.assertBucket(name);
+    if(this.bucket.toggles[name].variants[variantName]) {
+      this.bucket.toggles[name].variants[variantName]+=inc 
+    } else {
+      this.bucket.toggles[name].variants[variantName] = inc;
+    }
+    return true;
+  }
+
+  private bucketIsEmpty() {
     return Object.keys(this.bucket.toggles).length === 0;
   }
 
-  private resetBucket() {
-    const bucket: Bucket = {
+  private createBucket(): Bucket {
+    return {
       start: new Date(),
-      stop: null,
+      stop: undefined,
       toggles: {},
     };
-    this.bucket = bucket;
-  }
-
-  private closeBucket() {
-    if (this.bucket) {
-      this.bucket.stop = new Date();
-    }
   }
 
-  private getPayload(): Data {
-    this.closeBucket();
-    const payload = this.getMetricsData();
-    this.resetBucket();
-    return payload;
+  private resetBucket(): Bucket {
```

Comment:
My gut feel on this function is that it should either mutate state or return a value but not both, it becomes hard to reason about when you see the return value but it's also quietly mutating state behind the scenes

Reply:
yes, thus ended up as a foot gun. fixing.